presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSample.java

```diff
@@ -0,0 +1,293 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.ArrayType;
+import com.facebook.presto.common.type.Type;
+import io.airlift.slice.SizeOf;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.max;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ReservoirSample
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SingleReservoirSampleState.class).instanceSize();
+    private final Type type;
+
+    public Type getArrayType()
+    {
+        return arrayType;
+    }
+
+    private final Type arrayType;
+    /**
+     * Represents the list of sampled values.
+     * <br>
+     * We use an {@link ArrayList} of {@link Block} because the algorithm for
+     * the reservoir samples requires shuffling of elements. The {@link Block}
+     * interface doesn't have any method for setting values at arbitrary
+     * positions, so we resort to internally representing the sample as a list
+     * and then combining the samples into a single block later.
+     */
+    private ArrayList<Block> samples;
+    private int maxSampleSize = -1;
+    private long processedCount;
+
+    private Block initialSample;
+
+    public Block getInitialSample()
+    {
+        return initialSample;
+    }
+
+    public long getInitialProcessedCount()
+    {
+        return initialProcessedCount;
+    }
+
+    private long initialProcessedCount = -1;
+
+    public ReservoirSample(Type type)
+    {
+        this.type = requireNonNull(type, "type is null");
+        this.arrayType = new ArrayType(type);
+        this.samples = new ArrayList<>();
+    }
+
+    public ReservoirSample(ReservoirSample other)
+    {
+        this.type = other.type;
+        this.arrayType = other.arrayType;
+        this.processedCount = other.processedCount;
+        this.samples = other.samples;
+        this.maxSampleSize = other.maxSampleSize;
+        this.initialSample = other.initialSample;
+        this.initialProcessedCount = other.initialProcessedCount;
+    }
+
+    protected ReservoirSample(Type type, long processedCount, int maxSampleSize, ArrayList<Block> samples, Block initialSample, long initialSeenCount)
```

Comment:
re: checkArgument -- we can't do this easily because this constructor is used by the serializer to reconstruct the serialized state. It is very common for some of the state to be serialized before it's even initialized with a sample (e.g. maxSampleSize set to -1). The constructor isn't intended to be used outside the package, hence the `protected` modifier. I think it's ok to leave out the checks since outside usage is not intended.

I agree, having a `Block` argument type and reconstructing internally makes more sense. I will update it

Reply:
updated